### DISCUSSION TITLE
[Question][Lecture07] Validating the Weakest Liberal Precondition of a while loop

### DISCUSSION THREAD

**Discussion by jaehyun1ee** [prosyslab-classroom/cs424-program-reasoning#309] [Q&A]
I have one question regarding the weakest liberal precondition of a while loop.

In the class, we have learned that 
```
# given that I is the inductive invariant of the loop,

wlp(while E do S, Q) = I /\ (I /\ E -> wlp(I, S)) /\ (I /\ ㄱE -> Q)
```

My question is, **How can we show that the wlp is valid using Hoare rules of while?**

I tried in various ways, but failed to come up with a proper proof.

It is relatively easy to apply the Hoare Rule of if statement to its wlp. However, it was confusing how to apply the Hoare Rule of while to the wlp.

The Hoare rule of while is,

```
# given that I is the inductive invariant of the loop,

{I /\ E} S {I}
-------------------------
{I} while E do S {I /\ ㄱE}
```

Both the wlp and Hoare Rule builds up on the inductive invariant, `I`, which is given by some oracle. So this makes difficult in applying the rule for validating our wlp.

Put again, how can we prove that, 
```
{wlp(while E do S, Q)} while E do S {Q}
```
is valid using the Hoare Rule?

Sorry for a late question, and Thank you in advance :)

**Comment by KihongHeo** [comment]
Did you take a look at the updated slides? page 29. The old version was wrong.
Try with the new one.

**Comment by jaehyun1ee** [comment]
Oh, I didn't notice the change. I will try with the new version.

I have some follow-up questions though.

1. Then can I prove the validity by replacing `I` with `wlp(while E do S, Q)` in the following Hoare Rule,
```
# given that I is the inductive invariant of the loop,

{I /\ E} S {I}
-------------------------
{I} while E do S {I /\ ㄱE}

```

2. Then, are we showing that `wlp(while E do S, Q)` is the inductive invariant of the loop `while E do S`? What happens to the inductive invariant that we had in deriving `wlp(while E do S, Q)`? Are they both inductive invariants then?

**Comment by KihongHeo** [comment]
Hi,
- Relationship between Hoare rule and Wlp: This is always valid $\lbrace wlp(S,Q)\rbrace S  \lbrace Q \rbrace$
- `wlp(while E do S, Q)` is also an inductive invariant. Given $I$, the `wlp`of loop says that 1) loop invariant holds initially, 2) it holds after each iteration, and 3) it entails the postcondition.

**Comment by jaehyun1ee** [comment]
Thank you for the clarification, professor :)
